LIHADOOP-20318: Make Dr. Elephant analyze running jobs.

akshayrai opened the pull request:
This patch will support analysis of Mapreduce jobs in realtime soon after a task completes.

**Logic flow for better understanding this patch:**
- All the Completed jobs (succeeded and failed) fetched from the RM will go to a completed queue and will be analyzed by a set of threads(say, completedThreads) dedicated to them.
- All the Running jobs(jobs in undefined state) fetched from RM will go to a running queue and will be analyzed by another set of dedicated threads(say, runningThreads) which will compute the live metrics and potentially any new realtime heuristics.
- When a running job completes, it will be added to the completed queue and this will then be processed by a completedThread which will compute the final metrics and run all the heuristics.

**Code Changes:**
1. You can find the main logic implemented in `ElephanRunner.java, AnalyticJobGeneratorHadoop2.java and MapReduceFetcherHadoop2.java`
2. Added a GeneralConf.xml to configure the number of executor threads, realtime threads, fetch interval, fetch lag and real time update interval.
3. Added separate evolutions for h2(test) and mysql as they follow different syntax. 

**Schema changes**
1. Added a new level of severity called `Undefined` set to 0. We need to shift all the severities one level up as they currently start from 0.
2. Added a new field called status to yarn_app_result which tracks the status of the application. 

**Testing**
1. Added new tests.
2. Tested on local deployment.
3. Tested on a dev cluster.
4. Need to test on a live cluster. (TODO)

krishnap commented:
@akshayrai this will be super useful. haven't looked through the PR carefully yet. but we'll follow up with comments.

cc @paulbramsen @@plypaul @hongbozeng @liyintang @brandtg @timyitong

akshayrai commented:
Thanks @krishnap , glad you noticed the PR. I was about to send a request in the mailing group.

+ @shankar37 @nntnag17

nntnag17 commented:
Greate feature!

Some comments:

1. We should check the load that this feature will put on the Application master. I know it's very difficult to verify the load on Application master since it keeps changing but we need to check it before deployment.
2. Similarly we should also check if there are any network issues on Application master which may disturb the job execution. 
3. Will there be any performance issues on the database side since we are now doing multiple updates in the database. 
4. For running jobs, we need to find out the optimal number of executors and the fetch interval. There might be 100s of jobs running at a time, if the queue size for running jobs grow large and the rate of processing these jobs is not as fast, the jobs which are added at the back of the queue might be completed so dr. elephant might miss out on showing some of the running jobs at any given time. It might already be handled but just a concern. 


It would be great if you can share a screenshot of how the job summary looks for a running job.

nntnag17 commented on .gitignore:
```diff
@@ -17,9 +17,10 @@
-tmp
-dump
 .history
+.DS_Store
```
I see app/.DS_Store being checked in. Should that be included here as well?

akshayrai replied:
Ah! this should work. I must have checked in ```.DS_Store``` before adding it to gitignore. I'll remove the ```app/.DS_Store``` file

nntnag17 commented on app-conf/GeneralConf.xml:
```diff
@@ -0,0 +1,46 @@
+    <description>Number of threads to analyse completed(succeeded/failed) jobs.</description>
+  </property>
+  <property>
+    <name>drelephant.analysis.realtime.thread.count</name>
```
We need to find optimal combination of drelephant.analysis.realtime.thread.count and drelephant.analysis.fetch.interval to capture all the current running jobs.

akshayrai replied:
Yes, I did make a note of that point. It should actually be between the ```realtime.thread.count``` and ```realtime.update.interval```

shankar37 replied:
In the long run, we should move to async model. Having manually configuring the number of threads and having the thread sleep waiting for interval is highly inefficient.

nntnag17 commented on app/com/linkedin/drelephant/ElephantRunner.java:
```diff
@@ -45,136 +48,377 @@
         }
       });
-    } catch (Exception e) {
+    } catch (IOException e) {
```
I think initialize() throws a Runtime exception which might not be caught here

akshayrai replied:
You are right! thanks for pointing it out

nntnag17 commented on app/com/linkedin/drelephant/ElephantRunner.java:
```diff
@@ -45,136 +48,377 @@
+   * @param defaultValue The default value to use when key is not found
+   * @return the extracted value
+   */
+  private long getLongFromConf(Configuration conf, String key, long defaultValue) {
```
It looks like a utility method which might be used in other classes.

akshayrai replied:
Well, I did think about moving it to a utility class. But if you carefully take a look, the exception contains a message specific to GeneralConf.xml. Now, we can pass the conf file name as a string, but then the class will have too many arguments. So I left it in this class itself. And no other class is currently using it.

akshayrai commented:
Thanks @nntnag17 for your feedback. I do have a plan on addressing all the points you mentioned as part of testing Dr. Elephant on the live cluster.

Here is a snapshot.
![image](https://cloud.githubusercontent.com/assets/4448437/16347589/2d631248-3a6b-11e6-9f14-aa5470a5e88f.png)

shankar37 commented on app/com/linkedin/drelephant/ElephantRunner.java:
```diff
@@ -45,136 +48,377 @@
+  private class CompletedExecutorThread implements Runnable {
     private int _threadId;
-    private BlockingQueue<AnalyticJob> _jobQueue;
+    private DelayQueue<AnalyticJob> _completedJobQueue;
```
Can you add some comments on why you switched to DelayQueue from blocking queue ? I can see you need the DelayQueue for running jobs. Why do you need it for Completed Jobs ?

akshayrai replied:
When a job completes, there is a delay/lag in the job data moving to the job history server from the NM. I included a comment in line 242, where the expiry time is updated for a completed job.

shankar37 commented on app/com/linkedin/drelephant/ElephantRunner.java:
```diff
@@ -45,136 +48,377 @@
         } catch (Exception e) {
           logger.error(e.getMessage());
           logger.error(ExceptionUtils.getStackTrace(e));
+          retryAndDrop(analyticJob);
```
retryOrDrop

akshayrai replied:
Makes sense. Will update it.

shankar37 commented on app/com/linkedin/drelephant/ElephantRunner.java:
```diff
@@ -45,136 +48,377 @@
+            }
+          } else {
+            // Should not reach here. We consider only SUCCEEDED and FAILED JOBs in _completedJobQueue
+            throw new RuntimeException(analyticJob.getAppId() + " is in " + jobState.name() + " state. This should not"
```
You will throw an exception for non-completed jobs here which will be caught below and then added to the retry list. This will keep happening till max-retries are exhaust. Why not just log a message and ignore the job.

akshayrai replied:
Here we are analyzing jobs from the completedQueue. The completedJobs Queue will never have a non-completed job. So the code should never reach this section of code. However, if it does, then we need to fix the issue. But when it happens, we just want to retry it 3 times and then drop it.

shankar37 commented on app/com/linkedin/drelephant/ElephantRunner.java:
```diff
@@ -45,136 +48,377 @@
+                + analyticJob.getAppType().getName() + " " + analyticJob.getAppId() + " :: " + jobState.name());
+            if (analyticJob.getAppType().getName().equalsIgnoreCase(SPARK_APP_TYPE)) {
+              // Ignore as we do not capture any metrics for Spark jobs now
+              delayedEnqueue(analyticJob, _undefinedJobQueue);
```
your comment says we should ignore spark jobs but then why are you adding it to the _undefinedJobQueue again ?

akshayrai replied:
We are now fetching jobs by a start time range. So when we fetch a spark job which is running, it moves to the runningJobQueue. We can move this to the completedJobQueue only after it has completed. So we check the status of the spark job after a update interval and retain it in the undefinedJobQueue as long as it is not completed.

shankar37 commented on app/com/linkedin/drelephant/ElephantRunner.java:
```diff
@@ -45,136 +48,377 @@
+              if (AppResult.find.byId(analyticJob.getAppId()) == null) {
+                result.save();
+              } else {
+                result.update();
```
why do an update here but in the completedqueue processing you are doing delete and save ? Can you add comments explaining the rationale ?

akshayrai replied:
Good point. I added the comments in the code where we actually perform delete and save. I did that to prevent optimistic locking. It was an easy fix. Here, in case of running jobs, we do not get any optimistic locking likely because we update only one table(yarn_app_result). We can use a delete and save here as well but I thought an update would be more meaningful as we continuously update running jobs using the API. The final update happens by fetching data from the job history server in which case we delete records captured by the api and save the record obtained from the history server.

shankar37 commented on app/com/linkedin/drelephant/ElephantRunner.java:
```diff
@@ -188,10 +432,24 @@ private void waitInterval() {
+
+  public boolean isKilled() {
+    boolean killed = true;
+    if (_completedJobPool != null && _undefinedJobPool != null) {
```
it should be ||

akshayrai replied:
Both the pools should be shutdown.

shankar37 commented on app/com/linkedin/drelephant/ElephantRunner.java:
```diff
@@ -188,10 +432,24 @@ private void waitInterval() {
+    if (_completedJobPool != null && _undefinedJobPool != null) {
+      return !_running.get() && _completedJobPool.isShutdown() && _undefinedJobPool.isShutdown();
     }
+    return killed;
```
variable killed seems unnecessary

akshayrai replied:
yes! I'll update it.

shankar37 commented on app/com/linkedin/drelephant/analysis/AnalyticJobGeneratorHadoop2.java:
```diff
@@ -122,40 +122,61 @@ public void configure(Configuration configuration)
     URL succeededAppsURL = new URL(new URL("http://" + _resourceManagerAddress), String.format(
-            "/ws/v1/cluster/apps?finalStatus=SUCCEEDED&finishedTimeBegin=%s&finishedTimeEnd=%s",
-            String.valueOf(_lastTime + 1), String.valueOf(_currentTime)));
+            "/ws/v1/cluster/apps?finalStatus=SUCCEEDED&startedTimeBegin=%s&startedTimeEnd=%s",
```
Can you explain why you changed it to startedTime in the comments ?

shankar37 replied:
Also, will this cause any issue in the first time we deploy with this change to cause us to miss some completed jobs( that started before the finished time of the last analysed job before the restart) ?

akshayrai replied:
Earlier, when we were analyzing only succeeded and failed jobs, it made sense to use the finishedTime. But now that we analyze running jobs as well, we need to capture them soon after they start. For running jobs the finished time returned by the resource manager api is 0.

To answer your second question, this should not cause any problem because when we restart Dr. Elephant, we fetch all the jobs available in the resource manager starting from start time = 0(epoch) and then check in our database, if the application is already analyzed. If no, then we add it to the queue else we drop it.

shankar37 commented on app/com/linkedin/drelephant/analysis/AnalyticJobGeneratorHadoop2.java:
```diff
@@ -198,21 +219,31 @@ private JsonNode readJsonNode(URL url)
-      // When called first time after launch, hit the DB and avoid duplicated analytic jobs that have been analyzed
-      // before.
-      if (_lastTime > 0 || (_lastTime == 0 && AppResult.find.byId(appId) == null)) {
+      if (lastTime > 0 || (lastTime == 0 && AppResult.find.byId(appId) == null) ||
```
Can you add comment explaining what this condition is looking for readability purpose ? Or better yet, create micro functions for each condition and use them instead.

akshayrai replied:
Sure

shankar37 commented on app/com/linkedin/drelephant/analysis/Severity.java:
```diff
@@ -23,20 +23,23 @@
  * The severities that you can use
  */
 public enum Severity {
+  @EnumValue("5")
```
wouldn't this be a breaking change ? You need to update all existing records to new severity values.

akshayrai replied:
I have updated the mysql evolution files. We just have to increment the current severity values in the db before deploying Dr. Elpehant.

shankar37 commented on app/com/linkedin/drelephant/mapreduce/MapReduceFetcherHadoop2.java:
```diff
@@ -54,36 +53,60 @@
+    } else if (analyticJob.getJobStatus().equals("RUNNING")) {
+      jobData = fetchRunningJobsData(analyticJob);
+    } else {
+      // Dr. Elephant doesn't support states other than SUCCEEDED, FAILED and UNDEFINED. Code shouldn't reach here.
```
did you mean SUCCEEDED, FAILED and RUNNING ? Otherwise, your check is wrong as it's checking for RUNNING

akshayrai replied:
Yes, I'll update the comment.

shankar37 commented:
General comments. 
1. I didn't see a way to turn off this feature ? you should add a config for that if it's not already there.
2. we should have some performance tests done and it's report included. Ideally, we should have the instrumentation for job processed/sec, processing time etc and compare it with the feature on and off. Since that is still a work in progress, let's crunch some numbers manually. I am particularly concerned about how much time it takes to process a running job and whether we can process all of them before the configured interval. If we can't, then we need to figure out a way to not build up a huge backlog.
3. Can you also get the code coverage numbers with and without this change ? Since we add so much code, I want to make sure we are not degrading code coverage.

thanks
shankar

paulbramsen commented:
Been testing this out. Pretty cool! One thing to note: the jobtracker URL is broken for running jobs (tested with Hadoop 2.5.0).

akshayrai commented:
@shankar37  @paulbramsen thanks for your feedback. I'll update the PR and let your guys know.

paulbramsen commented:
@akshayrai any updates on this?

akshayrai commented:
@paulbramsen , I have been trying to setup a test instance on our cluster to do the testing. I'll update the PR once I am done with the complete testing.

paulbramsen commented:
Great. Thanks @akshayrai

paulbramsen commented:
@akshayrai any idea on the timeline for this?

akshayrai commented:
@paulbramsen , lately I have been busy with some other work. This work got delayed because of the test setup. Now that our setup is ready, I'll work on it and update it by end of next week.

Just a heads up. I'll be visiting Mountain View in the 2nd week of August. We can plan for some meetup.

paulbramsen commented:
@akshayrai hey have you made any progress on this?
I'm thinking about trying to rebase/merge myself but don't want to duplicate work if you already did it.

paulbramsen commented:
@akshayrai ok. Could you push your rebased/merged version of the project? That way we could take a look as well.

akshayrai commented:
@paulbramsen , I haven't rebased it with the master yet, if that is what you meant.

saguziel commented:
The catch-all exception in ElephantRunner is probably not a good idea. Since we both add it to the successful queue and and consider retrying it, my completed job queue ends up being huge with many duplicates.

shankar37 commented on app/com/linkedin/drelephant/ElephantRunner.java:
```diff
@@ -188,10 +310,160 @@ private void waitInterval() {
     }
   }
 
+  private void stopFurtherAnalysis(String appId) {
```
stopRealtimeAnalysis seems to be a better name as it is used only for realTime analysis.

akshayrai replied:
Sure, I'll rename it.

shankar37 commented on app/com/linkedin/drelephant/ElephantRunner.java:
```diff
@@ -188,10 +310,160 @@ private void waitInterval() {
+        if (appState.equals(SUCCEEDED) || appState.equals(FAILED) || appState.equals(KILLED)) {
+          AppResult result = _analyticJob.getAnalysis(null);
+          stopFurtherAnalysis(_analyticJob.getAppId());
+          saveOrUpdate(result);
```
You need to ensure the current scheduled job in _undefinedThreadPoolExecutor for this appId is finished before saving the result from the CompletedJob. Otherwise, it might overwrite it. stopFurtherAnalysis doesn't seem to ensure that.

akshayrai commented on app-conf/compile.conf:
```diff
@@ -1,3 +1,3 @@
-hadoop_version=2.3.0
+hadoop_version=2.4.1
```
It is just a conf I used while testing. Shouldn't be a concern but anyways I'll revert it to 2.3.0.

akshayrai commented on app-conf/elephant.conf:
```diff
@@ -20,4 +20,4 @@ db_password=""
 # Additional Configuration
 # Check https://www.playframework.com/documentation/2.2.x/ProductionConfiguration
-jvm_args="-Devolutionplugin=enabled -DapplyEvolutions.default=true"
+jvm_args="-Devolutionplugin=enabled -DapplyEvolutions.default=true -jvm-debug 9999"
```
I think it is good to keep it. It will help developers easily open a port to debug Dr. Elephant. There shouldn't be a problem retaining it.

akshayrai commented on app/com/linkedin/drelephant/ElephantRunner.java:
```diff
@@ -127,54 +137,166 @@ public Void run() {
+      return;
+    }
+
+    logger.info("Completed Job queue size is " + (getCompletedThreadPoolQueueSize() + completedJobs.size()));
```
We do log them. Here is a sample from the logs.

```
Fetching all the analytic apps which started between 1474550048956 and 1474550108955
No of active completed threads: 5
No of active running threads: 5
Fetched 109 undefined applications.
Fetched 101 completed applications.
Completed Job queue size is 1542
Undefined Job queue size is 546
```

akshayrai commented on app-conf/GeneralConf.xml:
```diff
@@ -0,0 +1,46 @@
+
+<!-- General Configurations -->
+
+<configuration>
```
Yes, we can disable realtime analysis by making the number of real time threads(```drelephant.analysis.realtime.thread.count```) 0.

shankar37 commented on app/com/linkedin/drelephant/analysis/AnalyticJobGeneratorHadoop2.java:
```diff
@@ -122,40 +124,60 @@ public void configure(Configuration configuration)
     }
 
-    _lastTime = _currentTime;
+    logger.info("Fetched " + appList.size() + " completed applications.");
```
I think there is value in separating the succeeded and failed applications list size

shankar37 commented on app/com/linkedin/drelephant/analysis/AnalyticJobGeneratorHadoop2.java:
```diff
@@ -122,40 +124,60 @@ public void configure(Configuration configuration)
-    logger.info("The succeeded apps URL is " + succeededAppsURL);
-    List<AnalyticJob> succeededApps = readApps(succeededAppsURL);
+            String.valueOf(from), String.valueOf(to)));
+    //logger.info("The succeeded apps URL is " + succeededAppsURL);
```
why comment this out ? Keeping it is useful I think

shankar37 commented on app/com/linkedin/drelephant/mapreduce/MapReduceFetcherHadoop2.java:
```diff
@@ -133,8 +174,131 @@ public MapReduceApplicationData fetchData(AnalyticJob analyticJob) throws IOExce
+
+    // Capture the job configuration
+    Properties props = new Properties();
+    if (job.getConfiguration() != null) {
```
Since this doesn't change, you can get it once and keep it instead of fetching it every min
